### PR TITLE
[WIP] transformed incremental + Rotations & Permutations

### PR DESCRIPTION
**Pull request by TLESORT**

### CODE REVIEW

File: clloader/scenarios/permutations.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self, seed):
+        self.seed = seed
+        self.g_cpu = torch.Generator()
+
+    def __call__(self, x):
+        shape = list(x.shape)
+        x = x.view(-1)
+        # if seed is 0, no permutations
+        if self.seed != 0:
+            self.g_cpu.manual_seed(self.seed)
```
Comment (arthurdouillard) [inline clloader/scenarios/permutations.py:23]: Shouldn't the seed be set in the `__init__`?

File: clloader/scenarios/transformation_incremental.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import numpy as np
+
+from clloader.task_set import TaskSet
+from clloader.datasets import _ContinuumDataset
+from clloader.scenarios import InstanceIncremental
+
+from torchvision import transforms
+
+class TransformationIncremental(InstanceIncremental):
```
Comment (arthurdouillard) [inline clloader/scenarios/transformation_incremental.py:11]: I'm not very aware of the litterature. I though that they were cases of rotation/permutation in NC mode?

File: continuum/scenarios/base.py [hunk 3]
```diff
@@ -88,12 +78,10 @@ def __getitem__(self, task_index: Union[int, slice]):
...
         :return: A train PyTorch's Datasets.
         """
-        data = self._select_data_by_task(task_index)
-        return TaskSet(
-            *data,
-            self.train_trsf if self.train else self.test_trsf,
-            data_type=self.cl_dataset.data_type
-        )
+        train = self._select_data_by_task(task_index)
+        train_dataset = TaskSet(*train, self.trsf, data_type=self.cl_dataset.data_type)
```
Comment (arthurdouillard) [inline continuum/scenarios/base.py:82]: Return directly the task set.

File: continuum/task_set.py [hunk 4]
```diff
@@ -45,19 +47,20 @@ def add_memory(
         :param t_memory: The associated task ids. If not provided, they will be
                          defaulted to -1.
         """
-        self.x = np.concatenate((self.x, x_memory))
-        self.y = np.concatenate((self.y, y_memory))
+
```
Comment (arthurdouillard) [inline continuum/task_set.py:50]: No empty line after docstring.

File: continuum/task_set.py [hunk 5]
```diff
@@ -70,15 +73,36 @@ def plot(
 
     def __len__(self) -> int:
         """The amount of images in the current task."""
-        return self.x.shape[0]
+        return self._x.shape[0]
+
+    def get_samples_from_ind(self, indices):
```
Comment (arthurdouillard) [inline continuum/task_set.py:78]: We should agree whether we use `indices` or `indexes`. I've used the latter everywhere.

File: continuum/task_set.py [hunk 6]
```diff
@@ -45,19 +47,20 @@ def add_memory(
         :param t_memory: The associated task ids. If not provided, they will be
                          defaulted to -1.
         """
-        self.x = np.concatenate((self.x, x_memory))
-        self.y = np.concatenate((self.y, y_memory))
+
+        self._x = np.concatenate((self._x, x_memory))
+        self._y = np.concatenate((self._y, y_memory))
         if t_memory is not None:
             self.t = np.concatenate((self.t, t_memory))
```
Comment (arthurdouillard) [inline continuum/task_set.py:54]: `self._t`

File: continuum/task_set.py [hunk 7]
```diff
@@ -92,14 +116,17 @@ def get_sample(self, index: int) -> np.ndarray:
...
+        y = self._y[index]
+        t = self._t[index]
 
         if self.trsf is not None:
             img = self.trsf(img)
 
         return img, y, t
 
+    def get_image(self, index):
+        return self.__getitem__(index)
```
Comment (arthurdouillard) [inline continuum/task_set.py:128]: `self[index]`

File: clloader/scenarios/transformation_incremental.py (hunk 2 above)
Comment (TLESORT) [reply clloader/scenarios/transformation_incremental.py:11]: Yes, most of the time it is NC (I personally think that permutations and rotations are interesting in the NI scenario). I did not find a satisfying solution to have both easily but the goal is to have both

File: continuum/task_set.py (hunk 5 above)
Comment (TLESORT) [reply continuum/task_set.py:78]: ok

File: clloader/scenarios/transformation_incremental.py (hunk 2 above)
Comment (TLESORT) [reply clloader/scenarios/transformation_incremental.py:11]: I close this comment because I will open another PR with the renamed branch

File: clloader/scenarios/permutations.py (hunk 1 above)
Comment (TLESORT) [reply clloader/scenarios/permutations.py:23]: le code est un peu tiré par les cheveux mais c'est parce que j'ai pas acces à la taille des points de données avant le __ call __ ...

**Comment by TLESORT** [comment]
On peut fermer cette PR une fois que les conversations sont terminées, la vrai PR est celle la https://github.com/Continvvm/continuum/pull/29
(J'ai créé une autre PR parce que j'ai changé le nom de la branch et que j'ai pas réussi à le faire dans la PR)
